### PR TITLE
Master

### PR DESCRIPTION
**Pull request by amitmiran**
KeepAlive + passing callback to handlers

### CODE REVIEW

File: ibapi.js [hunk 1]
```diff
@@ -26,26 +26,30 @@ NodeIbapi.prototype = {
...
-        handler(message);
+          var handler = this.handlers[message.messageId];
+          return handler(message, next);
+      };
+      if (!this.client.isConnected()) {
+          message = {};
+          message.messageId = 'disconnected';
+          if (message.messageId in this.handlers) {
+              var handler = this.handlers[message.messageId];
+              return handler(message, next);
```
Comment (dchem) [inline ibapi.js]: @amitmiran @antonzy What's the purpose of doing it this way?
Forcing message handlers to have f(data, cb) signature makes for awkward design patterns when attempting to expand on node-ibapi functionalities with FRP modules like RxJS, Bacon.js, or Kefir.js.

File: ibapi.js [hunk 2]
```diff
@@ -26,26 +26,30 @@ NodeIbapi.prototype = {
...
-    var message = this.client.getInboundMsg();
-    if (message.messageId in this.handlers) {
-      var handler = this.handlers[message.messageId];
-      handler(message);
-    };
-    if (!this.client.isConnected()) {
-      message = {};
-      message.messageId = 'disconnected';
+      this.client.checkMessages();
+      this.client.processMsg();
```
Comment (dchem) [inline ibapi.js]: Please make indentations two spaces wide.

File: ibapi.js (hunk 1 above)
Comment (antonzy) [inline ibapi.js]: Reactive programming is mainly used in client side applications, where race & concurrency are common practice. On the other hand, server side asynchronous tasks management is commonly translated into Task-To-Callback flow, if it's not callbacks then its "Promises" or similar abstractions. This gives us more fine grained control about the asynchronous nature of our server-side program, for example, preventing "taking care of" a new message before we finished taking care of the previous one. This keeps our application much consistent and predictable.
That's exactly why I pass the 'next' function call under the responsibility of the message handlers.

Comment (dchem) [inline ibapi.js]: @antonzy When you consider that in the processMessage it is running the ```handler``` function inline as opposed to returning it, isn't the following:
```js
var handleTickPrice = function( data ) {
  // do something
  if (predicate) {
    // do something else
    return;
  }
  // do something more
};
```
functionally same as this:
```js
var handleTickPrice = function( data, cb ) {
  // do something
  if (predicate) {
    // do something else
    cb();
  };
  // do something more
  cb();
};
```
so where does it make it more flexible in terms of the granular control that you mentioned? Let me know if I have this terribly wrong.

Or do you foresee use cases where you might be making async calls from within the handlers?

Also, regarding FRP - I've committed an example in the feature/frp-example branch - atsExample.js - that's how I foresee using FRP modules with node-ibapi. It should work fine even if we change back the function signature requirement, but it's just awkward.

**Comment by antonzy** [comment]
"Or do you foresee use cases where you might be making async calls from within the handlers?"
Exactly. All of our handlers are doing async operations. We cannot allow concurrent message handling as the order of the received messages is critical.

**Comment by dchem** [comment]
Right, but the message handling is definitely done in order and one at a time, so my confusion is why callback is needed when only one message is handled at a time - handler is called synchronously from within the ```processMessage```. Furthermore, in my module, all that ```async.forever``` does is to schedule the loop, not make the functions being scheduled to become asynchronous necessarily. It just schedules a function to run in series indefinitely: https://github.com/caolan/async#forever

Also, ```processMessage``` pops one message at a time (from C++ side's ```getInboundMsg```), and won't continue to next message handling until ```handler(data)``` has run its course through. So there is no danger of multiple messages being handled at a time, unless you are making an async calls from within the message handler function (like promise, for an example).

So just to clarify, you foresee doing something like this:
```js
function qSomePromise(data) {
  return new Promise( function (resolve, reject) {
    var arg = doSomethingWith(data);
    someAsyncFunc(arg, function(err, res) {
      if (err) {
        reject(err);
      } else {
        resolve(res);
      }
    });
  });
}
var handleData = function (data, callback) {
  qSomePromise(data).then( function(res) {
    // do something with res
    callback();
  });
}
```

**Comment by antonzy** [comment]
The example code you provided is exactly how we work with ibapi.
We have multiple asynchronous workflows processing each message, including passing it out to IO (database persistence, publish on rabbitmq & etc)  and we must wait for them to finish successfully before we can start handling the next one.

**Comment by dchem** [comment]
@antonzy Ok, as a compromise, how about this?

https://github.com/dchem/node-ibapi-addon/commit/2e3c6bc75e8019b6379430a1915877b3f6f6c699

**Comment by dchem** [comment]
@antonzy @amitmiran Did you guys make a decision on whether you want my suggested changes?

https://github.com/dchem/node-ibapi-addon/commit/2e3c6bc75e8019b6379430a1915877b3f6f6c699

**Pull request closed without merging**